File: src/main/java/com/trilead/ssh2/crypto/keys/Ed25519KeyPairGenerator.java
package com.trilead.ssh2.crypto.keys;

import com.google.crypto.tink.subtle.Ed25519Sign;

import java.security.GeneralSecurityException;
import java.security.KeyPair;
import java.security.KeyPairGeneratorSpi;
import java.security.SecureRandom;

/**
 * KeyPairGenerator implementation for Ed25519 keys.
 * <p>
 * Generates Ed25519 key pairs for SSH authentication and signing.
 *
 * @see Ed25519PublicKey
 * @see Ed25519PrivateKey
 */
public class Ed25519KeyPairGenerator extends KeyPairGeneratorSpi {
	@Override
	public void initialize(int keySize, SecureRandom secureRandom) {
		// ignored.
	}

	@Override
	public KeyPair generateKeyPair() {
		try {
			Ed25519Sign.KeyPair kp = Ed25519Sign.KeyPair.newKeyPair();
			return new KeyPair(new Ed25519PublicKey(kp.getPublicKey()), new Ed25519PrivateKey(kp.getPrivateKey()));
		} catch (GeneralSecurityException e) {
			throw new IllegalStateException(e);
		}
	}
}

File: sshlib/src/main/kotlin/org/connectbot/sshlib/crypto/ed25519/Ed25519KeyPairGenerator.kt
/*
 * ConnectBot SSH Library
 * Copyright 2025 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.sshlib.crypto.ed25519

import com.google.crypto.tink.subtle.Ed25519Sign
import java.security.GeneralSecurityException
import java.security.KeyPair
import java.security.KeyPairGeneratorSpi
import java.security.SecureRandom

internal class Ed25519KeyPairGenerator : KeyPairGeneratorSpi() {
    override fun initialize(keySize: Int, secureRandom: SecureRandom) {
        // ignored
    }

    override fun generateKeyPair(): KeyPair {
        try {
            val kp = Ed25519Sign.KeyPair.newKeyPair()
            return KeyPair(Ed25519PublicKey(kp.publicKey), Ed25519PrivateKey(kp.privateKey))
        } catch (e: GeneralSecurityException) {
            throw IllegalStateException(e)
        }
    }
}
